File: tests/phase6c_delayed_retrigger.py
#!/usr/bin/env python3
"""Negative controls for Phase 6C delayed/retrigger observations."""
from __future__ import annotations

import importlib.util
import json
from pathlib import Path
import shutil
import subprocess
import tempfile
import unittest

ROOT = Path(__file__).resolve().parents[1]
SPEC = importlib.util.spec_from_file_location(
    "phase6c_delayed_retrigger",
    ROOT / "scripts/phase6c-delayed-retrigger-evidence.py",
)
m = importlib.util.module_from_spec(SPEC)
assert SPEC.loader is not None
SPEC.loader.exec_module(m)


def capture(offset, track, note):
    return {
        "offset": offset,
        "track": track,
        "note": note,
        "command": 0,
        "parameter": 0,
    }


def valid_value():
    beat, tick = m.expected_time()
    return {
        "schema_version": 1,
        "load_returned": True,
        "song_name": m.TITLE,
        "bpm": 137.0,
        "tick_speed": 8,
        "is_ticks": True,
        "sample_rate": 44100,
        "samples_per_beat": beat,
        "samples_per_tick": tick,
        "loaded_parameters": dict(m.LOADED_PARAMETERS),
        "marker_position_after_extended": 0.25,
        "note_delay_events": [
            capture(m.LOADED_PARAMETERS["note_delay"] / 256.0, 0, 48)
        ],
        "retrigger_events": [
            capture(offset, 1, 50) for offset in m.expected_retrigger_offsets()
        ],
        "retr_cont_events": [
            capture(offset, 2, 52) for offset in m.expected_retr_cont_offsets()
        ],
        "reports": ["Load Warning: " + m.WARNING],
    }


def clean_log():
    return (
        "log: 0us: T: thread-id-123: # universalis # ../src/universalis/os/thread_name.cpp:56 # void universalis::os::thread_name::set_tls()\n"
        "log: 1us: T: thread-id-123: setting name for thread: id: 123, name: delayed-retrigger-probe\n"
        "log: 2us: T: delayed-retrigger-probe: # psycle-core # ../src/psycle/core/player.cpp:66 # void psycle::core::Player::start_threads()\n"
        "log: 3us: T: delayed-retrigger-probe: psycle: core: player: starting scheduler threads\n"
        "log: 4us: I: delayed-retrigger-probe: psycle: core: player: using 1 threads\n"
        "log: 5us: T: delayed-retrigger-probe: psycle: core: psy3 loader: loading psycle song fileformat version 3: /tmp/" + m.FIXTURE + "\n"
        "log: 6us: W: delayed-retrigger-probe: " + m.WARNING + "\n"
        "log: 7us: I: delayed-retrigger-probe: psycle: core: machine factory: create machine: loading with host: 0, plugin: <sampler>\n"
        "log: 8us: I: delayed-retrigger-probe: psycle: core: machine factory: create machine: loading with host: 0, plugin: <master>\n"
        "log: 9us: T: thread-id-123: # psycle-core # ../src/psycle/core/player.cpp:452 # void psycle::core::Player::stop_threads()\n"
        "log: 10us: T: thread-id-123: terminating and joining scheduler threads ...\n"
        "log: 11us: T: thread-id-123: # psycle-core # ../src/psycle/core/player.cpp:454 # void psycle::core::Player::stop_threads()\n"
        "log: 12us: T: thread-id-123: scheduler threads were not running\n"
    ).encode()


class CandidateParser(unittest.TestCase):
    def parse(self, value=None, log=None, exit_code=0):
        if value is None:
            value = valid_value()
        return m.parse_probe(
            json.dumps(value).encode(),
            clean_log() if log is None else log,
            exit_code,
        )

    def test_clean_scheduling_observation(self):
        result = self.parse()
        self.assertEqual(result["observation"], "command-scheduling-observed")
        self.assertEqual(result["loaded_parameters"]["note_delay"], 15)
        self.assertEqual(result["marker_position_after_extended"], 0.25)

    def test_note_delay_load_conversion_is_frozen(self):
        value = valid_value()
        value["loaded_parameters"]["note_delay"] = 127
        self.assertEqual(self.parse(value)["observation"], "inconclusive")

    def test_note_delay_scheduled_offset_is_rederived(self):
        value = valid_value()
        value["note_delay_events"][0]["offset"] += 0.001
        self.assertEqual(self.parse(value)["observation"], "inconclusive")

    def test_retrigger_series_is_rederived(self):
        value = valid_value()
        value["retrigger_events"].pop()
        self.assertEqual(self.parse(value)["observation"], "inconclusive")

    def test_scheduled_tracks_are_bound_to_fixture_lanes(self):
        for field, expected_track in (
            ("note_delay_events", 0),
            ("retrigger_events", 1),
            ("retr_cont_events", 2),
        ):
            with self.subTest(field=field):
                value = valid_value()
                value[field][0]["track"] = expected_track + 99
                self.assertEqual(self.parse(value)["observation"], "inconclusive")

    def test_retr_cont_series_is_rederived(self):
        value = valid_value()
        value["retr_cont_events"][1]["offset"] += 0.001
        self.assertEqual(self.parse(value)["observation"], "inconclusive")

    def test_extended_lpb_changes_loaded_marker_geometry(self):
        value = valid_value()
        value["marker_position_after_extended"] = 0.125
        self.assertEqual(self.parse(value)["observation"], "inconclusive")

    def test_unknown_diagnostic_is_rejected(self):
        contaminated = clean_log() + (
            b"log: 13us: W: delayed-retrigger-probe: unexpected warning\n"
        )
        self.assertEqual(
            self.parse(log=contaminated)["observation"], "inconclusive"
        )

    def test_abnormal_exit_is_inconclusive(self):
        for code in (1, 65, 70, 124, None, False):
            with self.subTest(code=code):
                self.assertEqual(
                    self.parse(exit_code=code)["observation"], "inconclusive"
                )

    def test_artifact_path_cannot_escape(self):
        with tempfile.TemporaryDirectory() as temp:
            root = Path(temp)
            for value in ("../escape", "/tmp/escape", "a\\b"):
                with self.subTest(value=value), self.assertRaises(ValueError):
                    m.child(root, value)


class OriginalSourceContract(unittest.TestCase):
    def test_git_blob_is_sensitive_to_source_bytes(self):
        self.assertNotEqual(m.git_blob(b"a"), m.git_blob(b"b"))

    def test_source_receipt_identity_is_fail_closed(self):
        with tempfile.TemporaryDirectory() as temp:
            root = Path(temp)
            receipt = {
                "schema_version": 1,
                "phase": "6C",
                "scope": "original-source-observation",
                "contract": m.CONTRACT,
                "evidence_role": "original-source",
                "source_repository": "jpaquim/psycle",
                "source_commit": m.SOURCE_COMMIT,
                "command_ids": {
                    "note_delay": 0xFD,
                    "retrigger": 0xFB,
                    "retr_cont": 0xFA,
                    "extended": 0xFE,
                    "set_lpb_range": [0x00, 0x1F],
                },
                "semantics": {
                    "note_delay_counter": "((parameter+1)*SamplesPerRow())/256",
                    "retrigger_rate": "parameter+1",
                    "retr_cont_rate_override": "parameter high nibble when nonzero",
                    "extended_lpb": "FE00..FE1F calls SetBPM(-1, parameter)",
                },
                "files": {
                    "psycle/src/psycle/host/Player.cpp": {
                        "git_blob": m.PLAYER_BLOB,
                    },
                    "psycle/src/psycle/host/SongStructs.hpp": {
                        "git_blob": m.SONGSTRUCTS_BLOB,
                    },
                },
                "original_psycle_executed": False,
                "parity_status": "UNKNOWN",
            }
            (root / "original-source-delayed-retrigger.json").write_text(
                json.dumps(receipt)
            )
            self.assertEqual(
                m.validate_source(root)["source_commit"], m.SOURCE_COMMIT
            )
            receipt["files"]["psycle/src/psycle/host/Player.cpp"]["sha256"] = "0" * 64
            (root / "original-source-delayed-retrigger.json").write_text(
                json.dumps(receipt)
            )
            with self.assertRaises(ValueError):
                m.validate_source(root)
            receipt["files"]["psycle/src/psycle/host/Player.cpp"].pop("sha256")
            receipt["files"]["psycle/src/psycle/host/Player.cpp"]["git_blob"] = "0" * 40
            (root / "original-source-delayed-retrigger.json").write_text(
                json.dumps(receipt)
            )
            with self.assertRaises(ValueError):
                m.validate_source(root)
            receipt["files"]["psycle/src/psycle/host/Player.cpp"]["git_blob"] = m.PLAYER_BLOB
            receipt["command_ids"]["note_delay"] = 0
            (root / "original-source-delayed-retrigger.json").write_text(
                json.dumps(receipt)
            )
            with self.assertRaises(ValueError):
                m.validate_source(root)


class MatrixDelayedRetriggerComparison(unittest.TestCase):
    def setUp(self):
        self.temp = tempfile.TemporaryDirectory()
        self.addCleanup(self.temp.cleanup)
        self.root = Path(self.temp.name)
        shutil.copytree(ROOT / "phase6c", self.root / "phase6c")
        (self.root / "scripts").mkdir()
        for name in ("phase6c-validate-matrix.py", "phase6c-candidate-parse.py"):
            shutil.copyfile(ROOT / "scripts" / name, self.root / "scripts" / name)
        self.evidence = (
            self.root / "phase6c/evidence/sequencer-delayed-retrigger"
        )

    def check(self, success):
        result = subprocess.run(
            ["python3", "scripts/phase6c-validate-matrix.py"],
            cwd=self.root,
            capture_output=True,
            text=True,
            check=False,
        )
        self.assertEqual(
            result.returncode == 0,
            success,
            result.stdout + result.stderr,
        )

    def mutate(self, name, change):
        path = self.evidence / name
        value = json.loads(path.read_text())
        change(value)
        path.write_text(json.dumps(value))

    def delayed_row(self):
        path = self.root / "phase6c/compatibility-matrix.json"
        matrix = json.loads(path.read_text())
        row = next(
            item
            for item in matrix["contracts"]
            if item["id"] == "sequencer-delayed-retrigger"
        )
        return path, matrix, row

    def test_versioned_deferred_comparison_validates(self):
        self.check(True)

    def test_original_runtime_trace_cannot_be_invented(self):
        self.mutate(
            "original-delayed-retrigger.json",
            lambda d: d.update(runtime_execution_trace="observed"),
        )
        self.check(False)

    def test_source_inspection_cannot_be_relabelled_execution(self):
        self.mutate(
            "original-source-delayed-retrigger.json",
            lambda d: d.update(original_psycle_executed=True),
        )
        self.check(False)

    def test_candidate_schedule_is_frozen(self):
        self.mutate(
            "candidate-delayed-retrigger.json",
            lambda d: d["note_delay_events"][0].update(track=99),
        )
        self.check(False)

    def test_final_run_attribution_is_frozen(self):
        self.mutate(
            "comparison.json",
            lambda d: d["source_evidence"].update(workflow_run_id=1),
        )
        self.check(False)

    def test_original_liveness_projection_is_frozen(self):
        self.mutate(
            "original-delayed-retrigger.json",
            lambda d: d.update(process_running_before_termination=False),
        )
        self.check(False)

    def test_candidate_timing_context_is_frozen(self):
        self.mutate(
            "candidate-delayed-retrigger.json",
            lambda d: d.update(tick_speed=24),
        )
        self.check(False)

    def test_original_source_artifact_digest_is_frozen(self):
        self.mutate(
            "original-source-delayed-retrigger.json",
            lambda d: d["source_evidence"].update(artifact_digest="sha256:" + "0" * 64),
        )
        self.check(False)

    def test_projection_procedure_digest_is_frozen(self):
        self.mutate(
            "candidate-delayed-retrigger.json",
            lambda d: d["source_evidence"].update(procedure_sha256="0" * 64),
        )
        self.check(False)

    def test_deferred_verdict_cannot_be_relabelled_pass(self):
        self.mutate(
            "comparison.json",
            lambda d: d.update(verdict="PASS", classification_allowed=True),
        )
        self.check(False)

    def test_matrix_status_cannot_promote_without_new_observer(self):
        path, matrix, row = self.delayed_row()
        row["status"] = "PASS"
        path.write_text(json.dumps(matrix))
        self.check(False)


class GeneratedObserver(unittest.TestCase):
    def generate(self, source: Path, output: Path):
        return subprocess.run(
            [
                "python3",
                str(ROOT / "scripts/phase6c-build-delayed-retrigger-observer.py"),
                str(source),
                str(output),
            ],
            cwd=ROOT,
            capture_output=True,
            text=True,
            check=False,
        )

    def test_generated_observer_adds_only_explicit_delayed_fixture(self):
        with tempfile.TemporaryDirectory() as temp:
            output = Path(temp) / "generated.ps1"
            process = self.generate(
                ROOT / "scripts/phase6c-original-windows-fixtures-v2.ps1",
                output,
            )
            self.assertEqual(process.returncode, 0, process.stderr)
            text = output.read_text(encoding="utf-8")
            self.assertIn("[switch]$ObserveDelayedRetrigger", text)
            self.assertIn('name = "delayed-retrigger"', text)
            self.assertIn(
                'expected_contract = "sequencer-delayed-retrigger"', text
            )
            self.assertIn(
                'if (-not $process.HasExited -and [bool]$renderOne.dialog_closed) {',
                text,
            )

    def test_builder_accepts_crlf_checkout_of_pinned_base(self):
        with tempfile.TemporaryDirectory() as temp:
            temp_path = Path(temp)
            source = temp_path / "base.ps1"
            output = temp_path / "generated.ps1"
            canonical = (
                ROOT / "scripts/phase6c-original-windows-fixtures-v2.ps1"
            ).read_text(encoding="utf-8")
            source.write_bytes(canonical.replace("\n", "\r\n").encode())
            process = self.generate(source, output)
            self.assertEqual(process.returncode, 0, process.stderr)


if __name__ == "__main__":
    unittest.main()

File: tests/phase6c_delayed_retrigger.cpp
// SPDX-License-Identifier: GPL-2.0-or-later
// Project-authored observation harness for delayed/retrigger command scheduling.
#include <psycle/core/detail/project.private.hpp>
#include <psycle/core/song.h>
#include <psycle/core/player.h>
#include <psycle/core/machinefactory.h>
#include <psycle/core/playertimeinfo.h>
#include <psycle/core/sampler.h>
#include <psycle/core/pattern.h>
#include <psycle/core/patternevent.h>
#include <psycle/core/commands.h>
#include <universalis/os/loggers.hpp>
#include <universalis/os/thread_name.hpp>

#define private public
#include <psycle/core/sequencer.h>
#undef private

#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

namespace {
std::vector<std::string> reports;

void report(const std::string& message, const std::string& title) {
    reports.push_back(title + ": " + message);
}

std::string quoted(const std::string& text) {
    std::ostringstream out;
    out << '"';
    for (unsigned char c : text) {
        if (c == '"' || c == '\\') out << '\\' << c;
        else if (c < 32) {
            out << "\\u" << std::hex << std::setw(4) << std::setfill('0')
                << int(c) << std::dec;
        } else out << c;
    }
    out << '"';
    return out.str();
}

struct Capture {
    double offset;
    int track;
    int note;
    int command;
    int parameter;
};

class RecordingSampler : public psycle::core::Sampler {
public:
    RecordingSampler(psycle::core::MachineCallbacks* callbacks, int id)
        : Sampler(callbacks, id) {}

    void AddEvent(
        double offset, int track,
        const psycle::core::PatternEvent& event) override {
        captures.push_back(Capture{
            offset, track, event.note(), event.command(), event.parameter()
        });
    }

    std::vector<Capture> captures;
};

struct LoadedEvent {
    double position;
    int track;
    psycle::core::PatternEvent event;
};

std::vector<LoadedEvent> loaded_events(psycle::core::CoreSong& song) {
    std::vector<LoadedEvent> result;
    psycle::core::Sequence& sequence = song.sequence();
    for (psycle::core::Sequence::iterator line_it = sequence.begin();
         line_it != sequence.end(); ++line_it) {
        psycle::core::SequenceLine* line = *line_it;
        for (psycle::core::SequenceLine::iterator entry_it = line->begin();
             entry_it != line->end(); ++entry_it) {
            psycle::core::Pattern& pattern = entry_it->second->pattern();
            if (pattern.id() < 0) continue;
            for (psycle::core::Pattern::iterator it = pattern.begin();
                 it != pattern.end(); ++it) {
                result.push_back(LoadedEvent{
                    it->first.first, it->first.second, it->second
                });
            }
            return result;
        }
    }
    return result;
}

void emit_captures(const std::vector<Capture>& captures) {
    std::cout << '[';
    for (std::size_t i = 0; i < captures.size(); ++i) {
        if (i) std::cout << ',';
        std::cout << "{\"offset\":" << captures[i].offset
                  << ",\"track\":" << captures[i].track
                  << ",\"note\":" << captures[i].note
                  << ",\"command\":" << captures[i].command
                  << ",\"parameter\":" << captures[i].parameter << '}';
    }
    std::cout << ']';
}
}

int main(int argc, char** argv) {
    if (argc != 2) return 64;

    try {
        using namespace universalis::os::loggers;
        static stream_logger diagnostics(std::cerr);
        multiplex_logger::singleton().add(diagnostics);
        universalis::os::thread_name thread_name("delayed-retrigger-probe");

        psycle::core::Player& player = psycle::core::Player::singleton();
        psycle::core::MachineFactory& factory =
            psycle::core::MachineFactory::getInstance();
        factory.Initialize(&player);

        int result = 0;
        {
            psycle::core::CoreSong song;
            player.song(song);
            song.report.connect(&report);
            if (!song.load(argv[1])) {
                std::cout << "{\"schema_version\":1,\"load_returned\":false}"
                          << std::endl;
                result = 65;
            } else {
                std::vector<LoadedEvent> events = loaded_events(song);
                RecordingSampler* recorder = new RecordingSampler(
                    factory.getCallbacks(), 0);
                recorder->Init();
                song.ReplaceMachine(recorder, 0);

                psycle::core::PlayerTimeInfo info;
                info.setSampleRate(44100);
                info.setBpm(song.bpm());
                info.setTicksSpeed(song.tick_speed(), song.is_ticks());

                psycle::core::Sequencer sequencer;
                sequencer.set_song(song);
                sequencer.set_time_info(info);
                sequencer.set_player(player);

                std::vector<Capture> note_delay, retrigger, retr_cont;
                double marker_position = -1.0;
                int loaded_note_delay_parameter = -1;
                int loaded_retrigger_parameter = -1;
                int loaded_retr_cont_parameter = -1;
                int loaded_extended_parameter = -1;

                for (std::size_t i = 0; i < events.size(); ++i) {
                    psycle::core::PatternEvent event = events[i].event;
                    if (event.command() == psycle::core::commandtypes::NOTE_DELAY) {
                        loaded_note_delay_parameter = event.parameter();
                        recorder->captures.clear();
                        sequencer.execute_notes(events[i].position, events[i].track, event);
                        note_delay = recorder->captures;
                    } else if (event.command() == psycle::core::commandtypes::RETRIGGER) {
                        loaded_retrigger_parameter = event.parameter();
                        recorder->captures.clear();
                        sequencer.execute_notes(events[i].position, events[i].track, event);
                        retrigger = recorder->captures;
                    } else if (event.command() == psycle::core::commandtypes::RETR_CONT) {
                        loaded_retr_cont_parameter = event.parameter();
                        recorder->captures.clear();
                        sequencer.execute_notes(events[i].position, events[i].track, event);
                        retr_cont = recorder->captures;
                    } else if (event.command() == psycle::core::commandtypes::EXTENDED) {
                        loaded_extended_parameter = event.parameter();
                    } else if (event.note() == 55 && events[i].track == 4) {
                        marker_position = events[i].position;
                    }
                }

                std::cout << std::setprecision(15)
                          << "{\"schema_version\":1"
                          << ",\"load_returned\":true"
                          << ",\"song_name\":" << quoted(song.name())
                          << ",\"bpm\":" << song.bpm()
                          << ",\"tick_speed\":" << song.tick_speed()
                          << ",\"is_ticks\":" << (song.is_ticks() ? "true" : "false")
                          << ",\"sample_rate\":44100"
                          << ",\"samples_per_beat\":" << info.samplesPerBeat()
                          << ",\"samples_per_tick\":" << info.samplesPerTick()
                          << ",\"loaded_parameters\":{"
                          << "\"note_delay\":" << loaded_note_delay_parameter
                          << ",\"retrigger\":" << loaded_retrigger_parameter
                          << ",\"retr_cont\":" << loaded_retr_cont_parameter
                          << ",\"extended_lpb\":" << loaded_extended_parameter
                          << "}"
                          << ",\"marker_position_after_extended\":" << marker_position
                          << ",\"note_delay_events\":";
                emit_captures(note_delay);
                std::cout << ",\"retrigger_events\":";
                emit_captures(retrigger);
                std::cout << ",\"retr_cont_events\":";
                emit_captures(retr_cont);
                std::cout << ",\"reports\":[";
                for (std::size_t i = 0; i < reports.size(); ++i) {
                    if (i) std::cout << ',';
                    std::cout << quoted(reports[i]);
                }
                std::cout << "]}" << std::endl;
            }
        }

        factory.Finalize();
        return result;
    } catch (const std::exception& error) {
        std::cerr << "delayed-retrigger probe exception: " << error.what() << std::endl;
        return 70;
    } catch (...) {
        std::cerr << "delayed-retrigger probe unknown exception" << std::endl;
        return 70;
    }
}
